src/main/scala/com/databricks/sparkdl/ImageUtils.scala

```diff
@@ -0,0 +1,144 @@
+ * Copyright 2017 Databricks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.databricks.sparkdl
+
+import java.awt.image.BufferedImage
+import java.awt.{Color, Image}
+
+import org.apache.spark.image.ImageSchema
+import org.apache.spark.sql.Row
+
+object ImageUtils {
+
+  /**
+   * Takes a Row image (spImage) and returns a Java BufferedImage. Currently supports 1 & 3
+   * channel images. If the image has 3 channels, we assume the channels are in BGR order.
+   *
+   * @param rowImage Image in spark.ml.image format.
+   * @return Java BGR BufferedImage.
+   */
+  private[sparkdl] def spImageToBufferedImage(rowImage: Row): BufferedImage = {
+    val height = ImageSchema.getHeight(rowImage)
+    val width = ImageSchema.getWidth(rowImage)
+    val channels = ImageSchema.getNChannels(rowImage)
+    val imageData = ImageSchema.getData(rowImage)
+    require(
+      imageData.length == height * width * channels,
+      s"""| Only one byte per channel is currently supported, got ${imageData.length} bytes for
+          | image of size ($height, $width, $channels).
+       """.stripMargin
+    )
+    val image = new BufferedImage(width, height, BufferedImage.TYPE_3BYTE_BGR)
+
+    var offset, h = 0
+    var r, g, b: Byte = 0
+    while (h < height) {
+      var w = 0
+      while (w < width) {
+        channels match {
+          case 1 =>
+            b = imageData(offset)
+            g = b
+            r = b
+          case 3 =>
+            b = imageData(offset)
+            g = imageData(offset + 1)
+            r = imageData(offset + 2)
+          case _ =>
+            require(false, s"`Channels` must be 1 or 3, got $channels.")
+        }
+
+        val color = new Color(r & 0xff, g & 0xff, b & 0xff)
+        image.setRGB(w, h, color.getRGB)
+        offset += channels
+        w += 1
+      }
+      h += 1
+    }
+    image
+  }
+
+
+  /**
+   * Takes a Java BufferedImage and returns a Row Image (spImage). The
+   *
+   * @param image Java BufferedImage.
+   * @return Row image in spark.ml.image format with 3 channels in BGR order.
+   */
+  private[sparkdl] def spImageFromBufferedImage(image: BufferedImage): Row = {
+    val channels = 3
+    val height = image.getHeight
+    val width = image.getWidth
+
+    val decoded = new Array[Byte](height * width * channels)
+    var offset, h = 0
+    while (h < height) {
+      var w = 0
+      while (w < width) {
+        val color = new Color(image.getRGB(w, h))
```

Comment:
definitely more readable. we can probably do some array magic with the array version of getRGB if we wanted to try optimizing.

Reply:
we could note that, or not, whatever.